chore: Files Recent Persistence E2E Test Plan (#1284)
hiero-ledger/hiero-block-node: docs/test-plan/persistence/files-recent-persistence-e2e-test-plan.md

>
> _**NOTE**: assume that before each test the Block-Node under test is expecting
> the next block to be received to be with number 0000000000001234567 which
> resolves to `/blocks/000/000/000/000/123/4/0000000000001234567.blk.zstd`. This
> is enough to illustrate the path resolution logic as well. Also, assume that
> the block is valid, has content and will successfully pass verification.
> Finally, assume that the above-mentioned block is NOT persisted before the
> start of each test._
>
> _**NOTE**: assume that we have a working verification because the
> `Files Recent Persistence` relies on blocks to be verified in order to execute
> any logic._
>
> _**NOTE**: assume that the Block-Node under test is configured to have the
> compression algorithm set to `ZStandard` so the file extension for a persisted
> block will be `.blk.zstd`_

|                            Test Case ID | Test Name                                                               | Implemented ( :white_check_mark: / :x: ) |
|----------------------------------------:|:------------------------------------------------------------------------|:----------------------------------------:|

> |                                         | [_**<br/>Happy Path Tests<br/><br/>**_](#happy-path-tests)              |                                          |

Reviewer: If you're seeing a `-`, then that got added somewhere.  `&nbsp;` is a space, and never renders as `-`.
Perhaps a quick huddle would help to resolve this?

> |                                         | [_**<br/>Happy Path Tests<br/><br/>**_](#happy-path-tests)              |                                          |

Author: ✅ I've found what the issue was, thanks for the rendering example, the diff shows it. Updated.

| [E2ETC_HP_FRP_0001](#E2ETC_HP_FRP_0001) | `Verify Acknowledgement Returned After Successful Persistence`          |                   :x:                    |
| [E2ETC_HP_FRP_0002](#E2ETC_HP_FRP_0002) | `Verify Persisted Block File Location`                                  |                   :x:                    |
| [E2ETC_HP_FRP_0003](#E2ETC_HP_FRP_0003) | `Verify Persisted Block File Content`                                   |                   :x:                    |
| [E2ETC_HP_FRP_0004](#E2ETC_HP_FRP_0004) | `Verify Persisted Block File Discoverable After Successful Persistence` |                   :x:                    |
| [E2ETC_HP_FRP_0005](#E2ETC_HP_FRP_0005) | `Verify Persisted Block File can be Overwritten`                        |                   :x:                    |
| [E2ETC_HP_FRP_0006](#E2ETC_HP_FRP_0006) | `Verify Persisted Block in Rapid Succession`                            |                   :x:                    |

> |                                         | [_**<br/>Error Tests<br/><br/>**_](#error-tests)                        |                                          |

Author: See above.

verified.

##### Requirements

The `Files Recent Persistence` is required to persist blocks. These blocks,
which are essentially files, are required to have the exact same binary content
as the original block streamed to the Block-Node and received by the
`Files Recent Persistence`.

##### Expected Behaviour

- It is expected that a regular file is written at the properly resolved
  location
- It is expected that the content of the file is the same as the original block
  streamed.

##### Preconditions

- A publisher that is able to stream to the Block-Node under test.

##### Input

- Valid block `0000000000001234567` is streamed as items.

##### Output

- Regular Readable File:
  `/blocks/000/000/000/000/123/4/0000000000001234567.blk.zstd` has the same
  binary content as the binary data received as block items for the specified
  block.

##### Other

N/A

---

#### E2ETC_HP_FRP_0004

##### Test Name

`Verify Persisted Block File Discoverable After Successful Persistence`

##### Scenario Description

`Files Recent Persistence` will persist a block after it has been received and
verified.

##### Requirements

> The `Files Recent Persistence` is required to persist blocks. These blocks are
> required to be discoverable through the Block-Node's public API after they have
> been successfully persisted. The blocks are required to be accessible
> (readable).

Reviewer: Wording suggestion:
```suggestion
The `Files Recent Persistence` MUST persist blocks.<br/>
Persisted blocks MUST be discoverable through the Block-Node's public API.<br/>
Persisted blocks MUST be readable and accessible through the Block-Node's
public API.<br/>
A persisted blocks MUST be discoverable after it has been fully persisted.<br/>
A persisted block MUST NOT be discoverable before the block is fully persisted.
```

Author: ✅ Thank you, added.

- A client that can call the public APIs to read the block (i.e. `getBlock`).
- Client receives not found if initially attempts to read the block under test
  (expected).

##### Input

- Valid block `0000000000001234567` is streamed as items.

##### Output

- Client is able to use the node's public API to read the persisted block
  (i.e. `getBlock`).

##### Other

N/A

---

#### E2ETC_HP_FRP_0005

##### Test Name

`Verify Persisted Block File can be Overwritten`

##### Scenario Description

`Files Recent Persistence` will persist a block after it has been received and
verified. If the received and verified block's resolved path already exists,
no matter what the reason is, it must be overwritten with the received block's
data. An Acknowledged block is deemed final and immutable (especially when the
next block is received, verified, persisted and acknowledged), but that is
outside the scope of the `Files Recent Persistence`. As long as data comes in to
the`Files Recent Persistence` it will be persisted at the resolved path. The
`Files Recent Persistence` is not a decision maker, it simply persists all data
it receives.

##### Requirements

> The `Files Recent Persistence` is required to persist blocks. If a file is
> present at the resolved location of the block, it is required to be truncated
> and overwritten with the received data. The `Files Recent Persistence` is not
> a decision maker, it is simply required to persists all data it receives.

Reviewer: Writing normative language is a bit of a lost skillset; unfortunately.
Suggestions:
```suggestion
The `Files Recent Persistence` SHALL persist blocks.<br/>
If a file is present at the resolved location of a block, the plugin SHALL
truncate and overwrite that file with the received data.<br/>
The `Files Recent Persistence` SHALL NOT decide which data to
retain, but SHALL overwrite any pre-existing file.<br/>
The plugin SHALL persist all data it receives via the messaging facility.
```

Author: ✅ Thanks for the wording suggestions, I've attempted to improve all `Requirements` sections to fit this style. I am trying to improve my wording skills in general, your feedback helps me a lot!

- If a file already exists at that path, it will be truncated and overwritten
  with the received data.

##### Preconditions

- A publisher that is able to stream to the Block-Node under test.
- A regular file exists at the resolved location of the block under test with
  some arbitrary test data.
- A client that can call the public API to read the block (i.e. `getBlock`).

##### Input

- Valid block `0000000000001234567` is streamed as items, waiting for
  Acknowledgement (ensure it is persisted).

##### Output

- Regular Readable File:
  `/blocks/000/000/000/000/123/4/0000000000001234567.blk.zstd` exists.
- The file's binary content is the same as the binary data received via the
  stream
- A client is able to read the persisted block through the public
  API (i.e. `getBlock`).

##### Other

N/A

---

#### E2ETC_HP_FRP_0006

##### Test Name

`Verify Persisted Block in Rapid Succession`

##### Scenario Description

`Files Recent Persistence` will persist a block after it has been received and
verified.

##### Requirements

The `Files Recent Persistence` is required to persist blocks.
The `Files Recent Persistence` is required to be able to persist blocks in rapid

> succession with no apparent issues.

Author: Thanks for the suggestion, I've added breaks in the `Requirements` sections. Indeed, statements are more readable like this.

##### Requirements

The `Files Recent Persistence` MUST persist blocks.<br/>

> Persisted blocks MUSt have the exact same binary content as the original block

Reviewer: Typo:
```suggestion
Persisted blocks MUST have the exact same binary content as the original block
```

Author: ✅ Thanks, updated.

##### Scenario Description

`Files Recent Persistence` will persist a block after it has been received and
verified.</br>

> IO failures CAN occur during the write process.

Reviewer: Very minor suggestion:
```suggestion
IO failures MAY occur during the write process.
```

##### Scenario Description

`Files Recent Persistence` will persist a block after it has been received and
verified.</br>

> IO failures CAN occur during the write process.</br>
> No files or data related to the current block MUST be present on the filesystem
> after an IO failure occurs during the write process.

Reviewer: Very minor suggestion:
```suggestion
IO failures MAY occur during the write process.</br>
Files and data related to the current block MUST NOT be present on the filesystem
if an IO failure occurs during the write process.
```
